### ISSUE TITLE
Word bigram perplexity interpretive spectrum with academic provenance

### ISSUE THREAD

**Issue by craigtrim** [craigtrim/pystylometry#62]
## Objective

Build an objective, research-backed interpretive spectrum for **word bigram perplexity** — where a text lands on the spectrum tells you something meaningful about the author's style. pystylometry already computes word bigram entropy and perplexity (`compute_word_bigram_entropy`); this issue is about giving those numbers *meaning*.

## What Word Bigram Perplexity Measures

Word bigram perplexity is rooted in Shannon's information theory. For a text of N words, the cross-entropy under a bigram model is:

```
H(W) = -(1/N) × Σ log₂ P(wᵢ | wᵢ₋₁)
```

Perplexity is the exponentiation of this entropy:

```
PP(W) = 2^H(W)
```

Perplexity is the **weighted average branching factor** — the effective number of equally likely next words at each position. A perplexity of 50 means the model is as uncertain as choosing uniformly among 50 possible next words.

Stylometrically, this captures **sequential lexical predictability** — how conventional or surprising an author's word-to-word transitions are. This is distinct from vocabulary richness (which measures *what* words are used); bigram perplexity measures *how* words are sequenced.

## The Interpretive Spectrum

Synthesized from Chen et al. (2017), Mohseni et al. (2022), Wu et al. (2024), and GPTZero findings:

### Very Low Perplexity (Highly Predictable)
- **What lands here**: AI-generated text (unedited LLM output), technical manuals, legal boilerplate, children's early readers, formulaic genre fiction with stock phrases, corporate mission statements
- **Why**: Extremely conventional word-to-word transitions. Heavy reliance on collocational clichés and prefabricated phrases
- **Stylistic implication**: Maximum conventionality, minimum authorial distinctiveness

### Low-Moderate Perplexity
- **What lands here**: Commercial genre fiction (romance, thriller, mystery), mass-market journalism, self-help books, popular science writing
- **Why**: Mostly predictable word sequences with occasional variation. Prose that prioritizes clarity and accessibility
- **Stylistic implication**: Accessible, genre-appropriate prose. Readability-optimized writing

### Moderate Perplexity
- **What lands here**: Quality long-form journalism, mainstream literary fiction, well-crafted nonfiction, essay collections
- **Why**: Balance of conventional phrasing and surprising word choices. Syntactic variety, broader vocabulary deployment
- **Stylistic implication**: The "sweet spot" — accessible yet distinctive. Where most accomplished professional writing falls

### High Perplexity
- **What lands here**: Ambitious literary fiction, canonical works, prose stylists (Nabokov, Morrison, DeLillo), complex essayists
- **Why**: Diverse vocabulary in less predictable combinations. Nominal density, creative metaphors, unusual collocations
- **Stylistic implication**: Informationally dense, stylistically distinctive prose. What Wu et al. (2024) call "strategies to achieve an informationally dense literary style"

### Very High Perplexity
- **What lands here**: Experimental literature, modernist/postmodernist prose (Joyce, Pynchon, Beckett), stream-of-consciousness, avant-garde poetry, surrealist text
- **Why**: Deliberately unconventional word sequences, neologisms, syntactic disruption, violation of expected word co-occurrence patterns
- **Stylistic implication**: Maximum authorial idiosyncrasy. What Mohseni et al. (2022) characterize as maximum sequential unpredictability

### Genre-Level Entropy (Chen et al. 2017, Freiburg-Brown corpus)

| Text Type | Mean Word-Form Entropy |
|-----------|----------------------|
| News | 8.313 |
| Fiction | 8.050 |
| General Prose | 7.949 |
| Academic | 7.841 |

POS entropy tells a different story: **fiction** has the highest POS entropy, reflecting greatest syntactic variety.

### Canonical vs. Non-Canonical Fiction (Mohseni et al. 2022)

Canonical texts exhibit significantly higher sequential unpredictability:
- Noun ApEn: 1.89 (canonical) vs. 1.83 (non-canonical)
- Verb ApEn: 1.75 vs. 1.70
- Pronoun ApEn: 1.74 vs. 1.68
- All p < 0.001
- ApEn achieved 77.3% balanced accuracy distinguishing canonical from non-canonical

### Canonical vs. Control (Wu et al. 2024)

GPT-family models assign consistently higher perplexity to canonical texts. Leading causes: "a more heavily nominal style, together with a more diverse vocabulary."

### Author-Level Complexity (Estevez-Rams et al. 2019)

- **Shakespeare**: Lowest entropy density, highest excess entropy — most *structured innovation*
- **Doyle**: Intermediate values
- **Abbott**: Highest entropy density, lowest excess entropy — more redundant, less innovative
- All texts lie along the same line in complexity-entropy space (slope -13.87, intercept 48.70)

### Diachronic Trend (Gonzalez-Torre et al. 2024)

Word entropy of American English has risen steadily since ~1900:
**social media > news/magazines > fiction/nonfiction books**

## Relationship to Other Stylometric Measures

| Measure | What It Captures | Overlap with Bigram Perplexity |
|---------|-----------------|-------------------------------|
| TTR / MTLD | Vocabulary richness (bag-of-words) | Partial — both sensitive to vocabulary diversity, but perplexity adds sequential ordering |
| Sentence length | Syntactic complexity (crude proxy) | Weak — short surprising sentences can have higher perplexity than long formulaic ones |
| Readability (Flesch etc.) | Word length + sentence length | Surprisingly weak (Trotta & Guarasci 2020) — different linguistic phenomena |
| Stress entropy | Phonological predictability | Orthogonal — different level of analysis |

**Conclusion**: Bigram perplexity captures **sequential lexical predictability** — partially correlated with but not redundant with TTR, MTLD, sentence length, or readability.

## Caveats and Confounds

1. **Vocabulary size**: Perplexity extremely sensitive to model vocabulary. Must be fixed and documented
2. **Text length**: Short texts (<500 words) produce unstable estimates
3. **Training corpus**: Determines what counts as "predictable." Must be balanced and documented
4. **Smoothing method**: Kneser-Ney preferred for n-gram models
5. **Genre/topic effects**: Bigram perplexity conflates authorial style with genre conventions and topic. Compare within genre for author-level insights
6. **OOV handling**: Texts with many proper nouns or neologisms inflate perplexity for non-stylistic reasons
7. **Tokenization**: Must be standardized (see Issue #56)

## Implementation Notes

- pystylometry already computes word bigram entropy/perplexity in `ngrams/`
- This issue is about adding an **interpretive layer** — mapping raw perplexity to the spectrum above
- Consider reporting: raw perplexity, normalized z-score vs. reference distribution, spectrum tier label
- Minimum text length threshold: flag results below 500 words
- Need calibration against a reference corpus of known texts spanning the spectrum

## References (MLA)

Barlas, Georgios, and Efstathios Stamatatos. "Cross-Domain Authorship Attribution Using Pre-trained Language Models." *Artificial Intelligence Applications and Innovations (AIAI 2020)*, Springer, 2020, pp. 255-266.

Bentz, Christian, et al. "The Entropy of Words — Learnability and Expressivity across More than 1000 Languages." *Entropy*, vol. 19, no. 6, 2017, article 275.

Brown, Peter F., et al. "An Estimate of an Upper Bound for the Entropy of English." *Computational Linguistics*, vol. 18, no. 1, 1992, pp. 31-40.

Chen, Heng, Haitao Liu, and Gabriel Altmann. "Entropy in Different Text Types." *Digital Scholarship in the Humanities*, vol. 32, no. 3, 2017, pp. 528-542.

Estevez-Rams, E., A. Mesa-Rodriguez, and D. Estevez-Moya. "Complexity-Entropy Analysis at Different Levels of Organisation in Written Language." *PLOS ONE*, vol. 14, no. 5, 2019, article e0214863.

Feldkamp, Ida Marie, et al. "Modeling Multilayered Complexity in Literary Texts." *Proceedings of NoDaLiDa/Baltic-HLT 2025*, 2025, pp. 142-158.

Fourkioti, Olga, et al. "Language Models and Fusion for Authorship Attribution." *Information Processing & Management*, vol. 56, no. 6, 2019, article 102061.

Genzel, Dmitriy, and Eugene Charniak. "Entropy Rate Constancy in Text." *Proceedings of the 40th Annual Meeting of the ACL*, 2002, pp. 199-206.

Gonzalez-Torre, Irene, Bartolo Luque, and Lucas Lacasa. "The Rising Entropy of English in the Attention Economy." *Communications Psychology*, vol. 2, 2024, article 117.

Juola, Patrick. "Authorship Attribution." *Foundations and Trends in Information Retrieval*, vol. 1, no. 3, 2008, pp. 233-334.

Juola, Patrick, and R. Harald Baayen. "A Controlled-Corpus Experiment in Authorship Identification by Cross-Entropy." *Digital Scholarship in the Humanities*, vol. 20, Supplement 1, 2005, pp. 59-67.

Lagutina, Ksenia, et al. "A Survey on Stylometric Text Features." *Proceedings of the 25th FRUCT*, IEEE, 2019.

Martinc, Matej, et al. "Detection of Changes in Literary Writing Style Using N-grams as Style Markers." *PLOS ONE*, vol. 17, no. 7, 2022, article e0267590.

Mohseni, Mahdi, Christoph Redies, and Volker Gast. "Approximate Entropy in Canonical and Non-Canonical Fiction." *Entropy*, vol. 24, no. 2, 2022, article 278.

Montemurro, Marcelo A., and Damián H. Zanette. "Universal Entropy of Word Ordering Across Linguistic Families." *PLOS ONE*, vol. 6, no. 5, 2011, article e19875.

Rivera-Solis, Luis, et al. "Attributing Authorship via the Perplexity of Authorial Language Models." *PLOS ONE*, vol. 20, no. 6, 2025, article e0327081.

Shannon, Claude E. "A Mathematical Theory of Communication." *Bell System Technical Journal*, vol. 27, no. 3, 1948, pp. 379-423.

Shannon, Claude E. "Prediction and Entropy of Printed English." *Bell System Technical Journal*, vol. 30, no. 1, 1951, pp. 50-64.

Stamatatos, Efstathios. "A Survey of Modern Authorship Attribution Methods." *JASIST*, vol. 60, no. 3, 2009, pp. 538-556.

Takahira, Ryosuke, Kumiko Tanaka-Ishii, and Lukasz Debowski. "Entropy Rate Estimates for Natural Language." *Entropy*, vol. 18, no. 10, 2016, article 364.

Trotta, Dario, and Raffaele Guarasci. "Is Neural Language Model Perplexity Related to Readability?" *CLiC-it 2020*, 2020.

Wu, Yaru, et al. "Perplexing Canon: A Study on GPT-Based Perplexity for Canonical and Non-Canonical Literary Works." *LaTeCH-CLfL 2024*, ACL, 2024, pp. 172-184.
